### PR TITLE
ruby: Support "binary" settings for Rubocop and Solargraph

### LINKED ISSUE #5109: When starting SolarGraph, try to respect project-specific installation
### Check for existing issues

- [X] Completed

### Is your feature request related to a problem?

If a user has multiple versions of the SolarGraph language server, Zed should use the one that's most appropriate for the current project.

### Describe the solution you'd like

From this [comment](https://github.com/zed-industries/feedback/issues/113#issuecomment-1322651896) by @attilagyorffy

I believe Zed should ideally follow the [Convention Over Configuration paradigm](https://en.wikipedia.org/wiki/Convention_over_configuration) and be smart about the above without the need of any configuration necessary. I believe a sensible priority (from higher to lower) would be the following:

* Use the solargraph bin stub if present (`./bin/solargraph`)
* Use the bundled version of solargraph if specified by the bundler context (see method described above)
* Use the matching solargraph in the `$PATH` (which should match the version of Ruby -> This is the current implementation if I understand correctly)
* Fail with a helpful error message, and with a link to a publicly available piece that helps setting things up

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by vitallium**
Hello, this pull request adds support for specifying and using the "binary" settings for Rubocop and Solargraph LSPs. AFAIK, Ruby LSP does not require the bundler context but that could be added later easily.

In Ruby world, like in Node.js world, almost all
projects rely on project specific packages (gems) and their versions. Solargraph and Rubocop gems are usually installed as project dependencies. Attempting to use global installation of them fail in most cases due to incompatible or missing dependencies (gems).

To avoid that, Ruby engineers have the `bundler`
gem that provides the `exec` command. This command executes the given command in the context of the bundle.

This pull request adds support for pulling the `binary` settings to use them in starting both LSPs. For instance, to start the Solargraph gem in the context of the bundler, the end user must configure the binary settings in the folder-specific settings file like so:

```json
{
  "lsp": {
    "solargraph": {
      "binary": {
        "path": "/Users/vslobodin/Development/festivatica/bin/rubocop"
      }
    }
  }
}
```

The `path` key must be an absolute path to the `binstub` of the `solargraph` gem. The same applies to the "rubocop" gem. Side note but it would be awesome to use Zed specific environment variables to make this a bit easier. For instance, we could use the `ZED_WORKTREE_ROOT` environment variable:

```json
{
  "lsp": {
    "solargraph": {
      "binary": {
        "path": "${ZED_WORKTREE_ROOT}/bin/rubocop"
      }
    }
  }
}
```

But this is out of the scope of this pull request. The code is a bit messy and repeatable in some places, I am happy to improve it here or later.

References:
- https://bundler.io/v2.4/man/bundle-exec.1.html
- https://solargraph.org/guides/troubleshooting
- https://bundler.io/v2.5/man/bundle-binstubs.1.html

This pull request is based on these two pull requests:

- https://github.com/zed-industries/zed/pull/14655
- https://github.com/zed-industries/zed/issues/15001

Closes https://github.com/zed-industries/zed/issues/5109.

Release Notes:

- N/A

### CODE REVIEW

File: extensions/ruby/src/language_servers/rubocop.rs [hunk 1]
```diff
@@ -9,11 +11,43 @@ impl Rubocop {
...
+            if let Some(bin_args) = binary_settings.arguments {
+                args = Some(bin_args);
+            }
+        }
+        let command = if let Some(binary) = binary {
+            binary
+        } else {
+            self.language_server_binary_path(worktree)?
+        };
+        let args = args.unwrap_or_else(|| vec!["--lsp".into()]);
```
Comment (vitallium) [inline extensions/ruby/src/language_servers/rubocop.rs:46]: My idea is to move this code into the `ruby.rs` file, use the passed `language_server_id` to pull the binary settings, and then ask the corresponding LSP for the correct arguments. But I failed to achieve that in this version. Perhaps I need another round.

**Comment by vitallium** [comment]
Added documentation about using folder-specific settings and binary settings.

**Comment by vitallium** [comment]
Rebased against the `main` branch.

**Comment by vitallium** [comment]
Okay, I think this pull request is ready for review.

**Review by maxdeviant — APPROVED** [review]
Thanks!

**Comment by vitallium** [comment]
> Thanks!

Thank you for making the Ruby extension better than it was! I use that as an opportunity to learn something new.

**Pull request merged**